### PR TITLE
[XPU][DeepSeekV4]Add Deepseek-v4 qnorm_rope_kv_insert kernel bf16 & fp8 SYCL support

### PR DESCRIPTION
**Pull request by jjmiao1**
# Add DeepSeek-V4 QNorm + RoPE + KV Insert SYCL Kernel

## Purpose

Add a SYCL implementation for the DeepSeek-V4 fused **QNorm + RoPE + KV Cache Insert** kernel, supporting both BF16 and FP8 output paths via a runtime `cache_dtype` parameter.

This kernel fuses four operations into a single pass for the KV projection stage of the MLA pipeline:

1. **RMSNorm (QNorm):** Per-head RMS normalization on the KV projection output
2. **RoPE rotation:** GPT-J style RoPE on the rope dimensions of the latent KV
3. **Quantization (FP8 path):** Per-block FP8 E4M3 quantization with power-of-2 scales
4. **KV Cache Insert:** Writes the processed result into the paged KV cache at the correct slot

**Commit:** `d5ebd30` — Add Deepseek-v4 qnorm_rope_kv_insert kernel bf16 & fp8 SYCL support

### Files Changed

| File | Change |
|------|--------|
| `csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp` | New SYCL kernel (520 lines), bf16 + fp8 unified |
| `csrc/xpu/ops.h` | Function declaration |
| `csrc/xpu/torch_bindings.cpp` | Op registration (`def` + `impl`) |
| `CMakeLists.txt` | Add source to build |
| `tests/test_deepseek_qnorm_rope_kv_insert.py` | Pytest with bf16/fp8 correctness + edge case tests |

### Design

- Grid: one subgroup (SG_SIZE=16) per `(token, head)`, processes full head_dim in a single pass
- BF16 path: normalize → RoPE → write bf16 to KV cache
- FP8 path: normalize → RoPE → per-128-block absmax → pow2 scale → FP8 E4M3 quantize → write to cache
- Runtime `cache_dtype` parameter selects BF16 or FP8 path
- Negative slot values are skipped (padding tokens)
- Registered as `torch.ops._xpu_C.deepseek_qnorm_rope_kv_fp8_insert`

## Test Result

---

### Unit Test — All PASSED

```bash
pytest tests/test_deepseek_qnorm_rope_kv_insert.py -v
```

- `test_bf16_correctness`: num_tokens ∈ {1, 65, 128} × seed ∈ {0, 42}
- `test_fp8_correctness`: num_tokens ∈ {1, 65, 128} × seed ∈ {0, 42}
- `test_negative_slot_skipped`: verifies padding tokens are not written
- `test_nonsequential_slots`: validates non-contiguous slot mapping

### Kernel Benchmark: SYCL vs Triton

Platform: Intel Arc Pro B60 (HBM peak bandwidth: 456 GB/s), Warmup 10 iters, Timed 50 iters

#### BF16 KV Cache Path

| num_tokens | triton(us) | sycl(us) | triton BW(GB/s) | sycl BW(GB/s) | sycl MBU | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 109.57 | 47.32 | 1.22 | 2.82 | 0.6% | **2.32x** |
| 2 | 105.87 | 48.96 | 2.52 | 5.45 | 1.2% | **2.16x** |
| 4 | 103.39 | 51.41 | 5.16 | 10.38 | 2.3% | **2.01x** |
| 8 | 105.48 | 50.12 | 10.12 | 21.29 | 4.7% | **2.10x** |
| 16 | 103.74 | 55.48 | 20.57 | 38.47 | 8.4% | **1.87x** |
| 32 | 105.52 | 55.70 | 40.45 | 76.63 | 16.8% | **1.89x** |
| 64 | 104.92 | 60.61 | 81.36 | 140.84 | 30.9% | **1.73x** |
| 128 | 105.22 | 55.45 | 162.26 | 307.91 | 67.5% | **1.90x** |
| 256 | 105.69 | 56.16 | 323.11 | 608.03 | >100%* | **1.88x** |
| 512 | 189.36 | 174.92 | 360.67 | 390.45 | 85.6% | **1.08x** |
| 1024 | 406.41 | 358.04 | 336.10 | 381.50 | 83.7% | **1.14x** |
| 2048 | 846.76 | 702.37 | 322.63 | 388.95 | 85.3% | **1.21x** |
| 4096 | 1666.56 | 1394.79 | 327.85 | 391.73 | 85.9% | **1.19x** |
| 8192 | 3491.36 | 2784.61 | 312.99 | 392.42 | 86.1% | **1.25x** |

**SYCL achieves 1.08x–2.32x speedup** over Triton. Large token regime: ~86% MBU.

#### FP8 KV Cache Path

| num_tokens | triton(us) | sycl(us) | triton BW(GB/s) | sycl BW(GB/s) | sycl MBU | speedup |
|---:|---:|---:|---:|---:|---:|---:|
| 1 | 148.56 | 54.86 | 0.89 | 2.42 | 0.5% | **2.71x** |
| 2 | 150.15 | 55.44 | 1.77 | 4.80 | 1.1% | **2.71x** |
| 4 | 152.87 | 54.75 | 3.48 | 9.71 | 2.1% | **2.79x** |
| 8 | 146.72 | 54.16 | 7.25 | 19.64 | 4.3% | **2.71x** |
| 16 | 142.60 | 53.07 | 14.92 | 40.08 | 8.8% | **2.69x** |
| 32 | 144.73 | 54.50 | 29.39 | 78.06 | 17.1% | **2.66x** |
| 64 | 139.68 | 49.96 | 60.92 | 170.31 | 37.3% | **2.80x** |
| 128 | 140.34 | 47.55 | 121.26 | 357.90 | 78.5% | **2.95x** |
| 256 | 139.77 | 50.62 | 243.52 | 672.43 | >100%* | **2.76x** |
| 512 | 191.62 | 172.64 | 355.24 | 394.29 | 86.5% | **1.11x** |
| 1024 | 430.03 | 356.43 | 316.59 | 381.97 | 83.8% | **1.21x** |
| 2048 | 890.32 | 700.74 | 305.83 | 388.57 | 85.2% | **1.27x** |
| 4096 | 1733.01 | 1390.35 | 314.23 | 391.68 | 85.9% | **1.25x** |
| 8192 | 3630.02 | 2767.07 | 300.04 | 393.61 | 86.3% | **1.31x** |

**SYCL achieves 1.11x–2.95x speedup** over Triton. Large token regime: ~84–86% MBU.

\* Small token counts show >100% MBU due to L2 cache hits (data fits in cache, not hitting HBM).

### End-to-End Model Benchmark (FP8 path)

**Integration:** In `vllm/models/deepseek_v4/xpu/xpu_qnorm_rope_kv_fp8_insert.py`, the SYCL path calls `torch.ops._xpu_C.deepseek_qnorm_rope_kv_insert` to replace the original Triton `xpu_qnorm_rope_kv_fp8_insert` implementation

Model: DeepSeek-V4-Flash-Base (**10 layers**, trimmed), TP=4, input_len=2048, output_len=1024, max_concurrency=1, num_prompts=5, num_warmups=2

| Metric | Triton (baseline) | SYCL | Delta |
|--------|---:|---:|---:|
| Output token throughput (tok/s) | 8.11 | 8.22 | +1.4% |
| Total token throughput (tok/s) | 24.32 | 24.67 | +1.4% |
| Mean TTFT (ms) | 2263.25 | 2362.58 | +4.4% |
| Median TTFT (ms) | 2269.85 | 2362.68 | +4.1% |
| P99 TTFT (ms) | 2276.82 | 2368.46 | +4.0% |
| Mean TPOT (ms) | 121.25 | 119.43 | **-1.5%** |
| Median TPOT (ms) | 121.28 | 119.38 | **-1.6%** |
| P99 TPOT (ms) | 121.74 | 119.92 | **-1.5%** |

### CODE REVIEW

**Comment by jjmiao1** [comment]
@jikunshang please help review, thanks

File: csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp [hunk 1]
```diff
@@ -0,0 +1,560 @@
...
+
+    // Q: sub-group reduction for RMSNorm.
+    float rrms = 1.0f;
+    auto sg = item.get_sub_group();
+    if (!is_kv) {
+      sum_sq = sycl::reduce_over_group(sg, sum_sq, sycl::plus<float>());
+      rrms = sycl::rsqrt(sum_sq / static_cast<float>(HEAD_DIM) + eps_);
+    }
+
+    if (is_kv) {
```
Comment (jikunshang) [inline csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp]: could L324, L329, L402 be merged?

File: csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp [hunk 2]
```diff
@@ -0,0 +1,560 @@
...
+  const int64_t q_stride_token = q.stride(0);
+  const int64_t q_stride_head = q.stride(1);
+  const int64_t kv_stride_token = kv.stride(0);
+  const int64_t cache_stride_pos = cos_sin_cache.stride(0);
+
+  constexpr int SG_SIZE = 16;
+  sycl::range<3> global(num_tokens, num_heads_q + 1, SG_SIZE);
+  sycl::range<3> local(1, 1, SG_SIZE);
+
+  auto& q_xpu = vllm::xpu::vllmGetQueue();
```
[the remark this answers is not in the document]
Comment (jikunshang) [reply csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp:510]: I feel this maybe necessary.

File: csrc/xpu/torch_bindings.cpp [hunk 3]
```diff
@@ -78,6 +78,21 @@ TORCH_LIBRARY_EXPAND(TORCH_EXTENSION_NAME, xpu_ops) {
       "apply_rotary_emb(Tensor! output, Tensor input,"
       "                 Tensor cos, Tensor sin, bool is_neox) -> ()");
   xpu_ops.impl("apply_rotary_emb", torch::kXPU, &apply_rotary_emb);
+  
+  // DeepSeek-V4 fused Q-RMSNorm + GPT-J RoPE + KV cache insert.
+  // kv_cache_dtype: "bf16" or "fp8_ds_mla"
+  xpu_ops.def(
+      "deepseek_qnorm_rope_kv_insert(Tensor! q, Tensor kv,"
+      "                              Tensor! cache,"
```
[the remark this answers is not in the document]
Comment (jikunshang) [reply csrc/xpu/torch_bindings.cpp:86]: PTAL.

**Comment by jikunshang** [comment]
please fix pre commit issue.

File: csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp [hunk 4]
```diff
@@ -0,0 +1,560 @@
...
+#include <sycl/sycl.hpp>
+#include "utils.h"
+#include <cstdint>
+#include <torch/all.h>
+
+namespace vllm {
+
+using bf16_t = sycl::ext::oneapi::bfloat16;
+
+static constexpr int HEAD_DIM = 512;
```
Comment (mayuyuace) [inline csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp]: So the kernel is only for deepseek-V4.
Is there possible that other models use this kernel, for example subsequent deepseek models?
If that, it is better to use template to dispatch these params instead of hard code.

Comment (jjmiao1) [reply csrc/xpu/sycl/deepseek_qnorm_rope_kv_insert.cpp]: Thanks for the comment, yes it is better to use template. Refactored to template `<int HEAD_DIM_, int ROPE_DIM_>` in the latest commit and currently instantiated with config <512, 64> for deepseek-v4.

**Review by Wei-Lin-Intel — APPROVED** [review]
LGTM

**Review by jikunshang — APPROVED** [review]

**Review by mayuyuace — APPROVED** [review]
